### PR TITLE
Add the new version of the contract

### PR DESCRIPTION
**Pull request by Daniel-K-Ivanov**
- Deprecates the old escrow v2 version
- Introduces new version of the escrow v3
- Adds transaction gas price into the authorisation signature and it is provided as parameter (used for msg sender refund)
-  Update the etherlime version to 1.1.0 in order to be able to perform `etherlime coverage`
- Add new unit tests for the relayed payment

### CODE REVIEW

File: contracts/Escrow_V2.sol [hunk 1]
```diff
@@ -1,9 +1,12 @@
 pragma solidity 0.5.1;
 
 import "./ECTools.sol";
-import "./../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
-import "./../node_modules/openzeppelin-solidity/contracts/token/ERC20/ERC20.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
```
Comment (bakasura980) [inline contracts/Escrow_V2.sol:4]: Not every editor supports it. In Visual Studio Code it does not find the path

File: contracts/Escrow_V3.sol [hunk 2]
```diff
@@ -0,0 +1,149 @@
+pragma solidity 0.5.1;
+
+import "./ECTools.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
```
Comment (bakasura980) [inline contracts/Escrow_V3.sol:4]: Not every editor supports it. In Visual Studio Code it does not find the path

File: test/escrow-tests.js [hunk 3]
```diff
@@ -363,4 +407,14 @@ describe('Escrow Contract', function () {
             );
         });
     });
-});
\ No newline at end of file
+});
+
+async function verifyContractBalanceAfterRefund(deployer, escrowContract, txReceipt, weiToSend) {
```
Comment (bakasura980) [inline test/escrow-tests.js]: Suggestion: `deployer` and `escrowContract` are  public variables for the global describe -> Move this function in the describe scope and remove `deployer, escrowContract` from function params. Also instead of `txReceipt`, provide `gasUsed` directly

File: test/escrow-tests.js [hunk 4]
```diff
@@ -93,106 +95,148 @@ describe('Escrow Contract', function () {
...
+        it('Should refund from fiat payment depending on gasPrice argument in authorisation signature', async () => {
+            const msgSenderBalanceBeforeFund = await deployer.provider.getBalance(dAppSigner.signer.address);
+
+            // Broadcast the transaction with x10 Gas Price. Should refund only 1x Gas Price and not 10x 
+            const tx = await escrowSignerExecutor.fundForFiatPayment(nonce, GAS_PRICE, recipient.address, weiToSend, tokensToSend, signedFiatPaymentFunds, { gasLimit: GAS_LIMIT, gasPrice: GAS_PRICE * 10 });
+            const txReceipt = await tx.wait();
+
+            const msgSenderBalanceAfterFund = await deployer.provider.getBalance(dAppSigner.signer.address);
+            // msg sender consumed a lot of ether on the fundForFiatPayment transaction, so the balance should be less then the initial balance. The refund must be only for Gas Price and not 10xGas Price
+            assert(msgSenderBalanceBeforeFund.gt(msgSenderBalanceAfterFund), "msg sender was refunded with more than necessary ethers");
```
Comment (bakasura980) [inline test/escrow-tests.js:223]: In the perfect case, we should refund the exact amount of the spent gas ? If so, add here gte

File: test/escrow-tests.js [hunk 5]
```diff
@@ -4,22 +4,24 @@ const etherlime = require('etherlime');
...
+const GAS_PRICE = 20000000000; // 20 gwei
+const GAS_LIMIT = 200000;
+
+const EXCESS_GAS_REFUND_UPPER_LIMIT = 3500; // The upper limit of exceeding gas that can be refunded and we will count the refund successfull
+const NUMBER_OF_TRANSACTIONS = 500; // Number of transactions that will be perfomed for the refund test
 
 describe('Escrow Contract', function () {
     this.timeout(5000);
-    let signer = accounts[3];
+    let dAppSigner = accounts[3];
```
Comment (bakasura980) [inline test/escrow-tests.js]: dAppSigner.signer is strange -> let dAppSigner = accounts[3].signer; ?

File: test/escrow-tests.js [hunk 6]
```diff
@@ -29,7 +31,7 @@ describe('Escrow Contract', function () {
     let escrowContract;
 
     async function initEscrowContract() {
-        deployer = new etherlime.EtherlimeGanacheDeployer(dAppAdmin.wallet.privateKey);
+        deployer = new etherlime.EtherlimeGanacheDeployer(dAppAdmin.signer.privateKey);
```
Comment (bakasura980) [inline test/escrow-tests.js:36]: Suggestion:

`
deployer = new etherlime.EtherlimeGanacheDeployer(dAppAdmin.signer.privateKey).signer;   
provider = deployer.provider;
`

Below you are using deployer.provider a lot. In my opinion it would be more cleaner if you just use provier.someMethod.

File: contracts/Escrow_V3.sol [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+        _refundMsgSender(gasLimit, gasprice);
+    }
+
+    /**
+    * @dev Funds the `addressToFund` with the proided `weiAmount` and `tokenAmount`
+    * Signature from the dapp is used in order to authorize the funding
+    * The msg sender is refunded for the transaction costs
+    */
+    function fundForFiatPayment(
```
Comment (bakasura980) [inline contracts/Escrow_V3.sol:76]: If the gas station goes down, we call this method with 1 gas price  by default, right ? If so, maybe this transaction won't be processed from miners. What would you say to set a MINIMUM_GAS_PRICE_LIMIT ?

File: test/escrow-tests.js [hunk 8]
```diff
@@ -4,22 +4,24 @@ const etherlime = require('etherlime');
...
 const Token = require('./../build/Token.json');
 const ECTools = require('./../build/ECTools.json');
-const EscrowContract = require('./../build/Escrow_V2.json');
+const EscrowContract = require('./../build/Escrow_V3.json');
 
-// TODO More tests must be implemented for the full coverage of the functionality and testing corner cases
+const GAS_PRICE = 20000000000; // 20 gwei
+const GAS_LIMIT = 200000;
+
+const EXCESS_GAS_REFUND_UPPER_LIMIT = 3500; // The upper limit of exceeding gas that can be refunded and we will count the refund successfull
```
Comment (bakasura980) [inline test/escrow-tests.js:12]: Where 3500 comes from ?

File: contracts/Escrow_V2.sol (hunk 1 above)
Comment (Daniel-K-Ivanov) [reply contracts/Escrow_V2.sol:4]: It results in a error but in order to use Etherlime coverage, the paths needs to be this way. Quantstamp's team recommended importing files this way as-well.

File: test/escrow-tests.js (hunk 3 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js]: It results in a error but in order to use Etherlime coverage, the paths needs to be this way. Quantstamp's team recommended importing files this way as-well.

File: contracts/Escrow_V3.sol (hunk 2 above)
Comment (Daniel-K-Ivanov) [reply contracts/Escrow_V3.sol:4]: It results in a error but in order to use Etherlime coverage, the paths needs to be this way. Quantstamp's team recommended importing files this way as-well.

File: test/escrow-tests.js (hunk 3 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js]: done

File: test/escrow-tests.js (hunk 4 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js:223]: You are right, but in this exact test case we are not doing that. 

In this unit test we are performing a fundFiat transaction but we are broadcasting the transaction with 10xGasPrice. The signer (dapp backend) signed the authorisation signature for 10GWEI gasprice, however we are broadcasting the transaction with 100 GWEI to check whether the Escrow will refund with 10GWEI and not 100GWEI.

That is why we are checking that the balance before fund is bigger than the balance after fund.

1. We broadcast a transaction for 100GWEI gas price
2. The msg sender must be refunded for 10GWEI not 100GWEI => 
3. The msg sender will have less balance after the transaction

File: test/escrow-tests.js (hunk 5 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js]: dAppSigner is the predefined wallet that is given from `ganache`.  I have named it dAppSigner because we have the term `signer` for the wallet that is signing the authorisation signatures. 

Unfortunately after etherlime 1.1.0 the account wallets are not reached with `account.wallet` but `account.signer` which is kind of confusing.

Good news is that every-time we use `dAppSigner` we are doing `dAppSigner.signer` so I can change the variable to `accounts[3].signer`

File: contracts/Escrow_V3.sol (hunk 7 above)
Comment (Daniel-K-Ivanov) [reply contracts/Escrow_V3.sol:76]: In my opinion it is better to handle the problem in the backend by using a more up-to-date version of the Gas station API.

The bare minimum gas price is 1 Gwei enforced by the protocol. If we set the minimum gas price to another value, what should this value be? 
If we set it to 5 gwei for example, we will be paying premium in gas price in times where the network has 1-5Gwei gasprice. 

Regarding the `transaction won't be processed from miners` - the transaction will be in the mempool unless some miner picks it up, so yeah it will take more time, but i think that we need to regulate this from the backend and not the contract

File: test/escrow-tests.js (hunk 8 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js:12]: This number is the gas amount that we can use as a buffer and still consider the refund successful.
In other words, if the transaction costs 100k GAS and we get 103K Gas refunded, we mark it as success, however if the refund is 104K gas the test will fail.

This number was used previously in the tests and it is an arbitrary number (gas buffer for the refunds)

File: test/escrow-tests.js (hunk 6 above)
Comment (Daniel-K-Ivanov) [reply test/escrow-tests.js:36]: will fix

**Review by bakasura980 — APPROVED** [review]
